hikari/impl/cache.py

```diff
@@ -1473,8 +1473,13 @@ def _garbage_collect_message(
             if guild_record:
                 self._garbage_collect_member(guild_record, message.object.member, decrement=1)
 
-        if message.object.referenced_message:
-            self._garbage_collect_message(message.object.referenced_message, decrement=1)
+        referenced_message = message.object.referenced_message
+        if not referenced_message and message.object.message_reference and message.object.message_reference.id:
+            msg_id = message.object.message_reference.id
+            referenced_message = self._message_entries.get(msg_id) or self._referenced_messages.get(msg_id)
+
```

Comment:
Also you're not accounting for what kind of reference this is so for stuff like cross posts, channel follows, and pins adds where you'd have message_reference.id but not referenced_message this could lead to other edge case bugs.